library/std/src/sys/pal/windows/fs.rs

```diff
@@ -1068,14 +1068,35 @@ pub fn readdir(p: &Path) -> io::Result<ReadDir> {
     unsafe {
         let mut wfd = mem::zeroed();
         let find_handle = c::FindFirstFileW(path.as_ptr(), &mut wfd);
+
+        // The status `ERROR_FILE_NOT_FOUND` is returned by the `FindFirstFileW` function
+        // if no matching files can be found, but not necessarily that the path to find the
+        // files in does not exist.
+        //
+        // Hence, a check for whether the path to search in exists is added when the last
+        // os error returned by Windows is `ERROR_FILE_NOT_FOUND` to handle this scenario.
+        // If that is the case, an empty `ReadDir` iterator is returned as it returns `None`
+        // in the initial `.next()` invocation because `ERROR_NO_MORE_FILES` would have been
+        // returned by the `FindNextFileW` function.
+        //
+        // See issue #120040: https://github.com/rust-lang/rust/issues/120040.
+        let last_error = Error::last_os_error();
```

Comment:
This needs to happen only if `find_handle == INVALID_HANDLE_VALUE`. It's unspecified what `GetLastError` returns if the function succeeds. While some functions do set it to `ERROR_SUCCESS` others don't and in either case it's not guaranteed.

Also you can use `api::get_last_error()` to get the error without needing to unwrap below. This can later be turned into an `std::io::Error` using `Error::from_raw_os_error(last_error.code as i32)`.